raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -584,211 +721,244 @@ private void updateVoterConnections(FindQuorumResponseData response) {
+        if (requestEpoch < quorum.epoch()) {
+            return Optional.of(Errors.FENCED_LEADER_EPOCH);
+        } else if (quorum.isObserver() || !quorum.isVoter(remoteNodeId)) {
+            // Inconsistent voter state could be the result of an active
+            // reassignment which has not been propagated to all of the
+            // expected voters. We generally expect this to be a transient
+            // error until all voters have replicated the reassignment record.
+            return Optional.of(Errors.INCONSISTENT_VOTER_SET);
         } else if (shutdown.get() != null) {
             shutdown.get().onEpochUpdate(requestEpoch);
-            return Optional.of(Errors.BROKER_NOT_AVAILABLE.exception());
+            return Optional.of(Errors.BROKER_NOT_AVAILABLE);
         }
 
-        // TODO: seems like a weird place to do this...
+        // The request is from a valid voter and indicates a larger epoch,
+        // so we transition to follower status before handling the request
         if (requestEpoch > quorum.epoch()) {
             becomeUnattachedFollower(requestEpoch);
         }
         return Optional.empty();
     }
 
-    private Optional<Exception> validateLeaderOnlyRequest(int requestEpoch) {
-        if (quorum.isObserver()) {
-            return Optional.of(new KafkaException("Observers are not allowed to receive requests"));
-        } else if (requestEpoch < quorum.epoch()) {
-            return Optional.of(Errors.FENCED_LEADER_EPOCH.exception());
-        } else if (requestEpoch > quorum.epoch()) {
-            // We cannot be the leader of an epoch we are not aware of
-            return Optional.of(Errors.UNKNOWN_LEADER_EPOCH.exception());
-        } else if (!quorum.isLeader()) {
-            return Optional.of(Errors.NOT_LEADER_FOR_PARTITION.exception());
+    /**
+     * Validate a request which is intended for the current quorum leader.
+     */
+    private Optional<Errors> validateLeaderOnlyRequest(int requestEpoch) {
+        if (requestEpoch < quorum.epoch()) {
+            return Optional.of(Errors.FENCED_LEADER_EPOCH);
+        } else if (requestEpoch > quorum.epoch() || !quorum.isLeader()) {
+            // If the epoch in the request is larger than our own epoch or
+            // it matches our epoch and we are not a leader, then the request
+            // is invalid. Unlike with Kafka topic partitions, leaders will
+            // always be the first to know of their status.
+            return Optional.of(Errors.INVALID_REQUEST);
         } else if (shutdown.get() != null) {
-            return Optional.of(Errors.BROKER_NOT_AVAILABLE.exception());
+            return Optional.of(Errors.BROKER_NOT_AVAILABLE);
         }
         return Optional.empty();
     }
 
     private void handleRequest(RaftRequest.Inbound request) throws IOException {
-        ApiMessage requestData = request.data();
+        ApiKeys apiKey = ApiKeys.forId(request.data.apiKey());
```

Comment:
nit: log4j.trace recording the recipient of the request?

Reply:
I have a trace message under `handleInboundMessage`.